syntax = "proto3";

package greet;

// The greeting service definition.
service Greeter {
    // Unary
    rpc SayHello (HelloRequest) returns (HelloReply);

    // Server streaming
    rpc ParrotSayHello (HelloRequest) returns (stream HelloReply);

    // Client streaming
    rpc ChattyClientSayHello (stream HelloRequest) returns (DelayedReply);

    // Bidirectional streaming
    rpc InteractiveSayHello (stream HelloRequest) returns (stream HelloReply);
}

// The request message containing the user's name.
message HelloRequest {
    string name = 1;
    string greeting = 2;
}

// The response message containing the greetings.
message HelloReply {
    string message = 1;    
}

// The response message containing delayed greetings.
message DelayedReply {
    string message = 1;
    repeated HelloRequest request = 2;
}
